src/box/xrow_update_array.c

```diff
@@ -448,10 +448,11 @@ xrow_update_op_do_array_##op_type(struct xrow_update_op *op,			\
 	}									\
 	if (item->field.type != XUPDATE_NOP)					\
 		return xrow_update_err_double(op);				\
-	if (xrow_update_op_do_##op_type(op, item->field.data) != 0)		\
+	const char *data = item->field.data;					\
+	xrow_update_mp_read_scalar(&data, &item->field.scalar);			\
```

Comment:
Is that correct that this commit (`update: keep result of scalar operation in tree itself`) makes the code parse the scalar operation argument multiple times (AFAIK, it was parsed only one time before).

Reply:
Sorry, I don't see how it is more then a single parse. Please, explain)

Reply:
You added here `xrow_update_mp_read_scalar()` which parses `item->field.data`. This code can be done by multiple update operations having the same path, right? Which means that every operation reaching this place will read the scalar field again. Wouldn't it?

Reply:
It is not quite so. After this patch the second update will still hit `item->field.type != XUPDATE_NOP` raising `xrow_update_err_double` error. And after the next patch which touches the same place the parse will be done only on the first time when `item->field.type == XUPDATE_NOP`. After that type is changed to `XUPDATE_SCALAR` and we only do the update operation itself without parsing.